### PR TITLE
[TRIVIAL] Improve reporting of missing node exceptions

### PR DESCRIPTION
**Pull request by nbougalis**

### CODE REVIEW

File: src/ripple/shamap/SHAMapMissingNode.h [hunk 1]
```diff
@@ -34,33 +36,44 @@ enum class SHAMapType
...
-    SHAMapType mType;
-    SHAMapHash mNodeHash;
-    uint256    mNodeID;
 public:
-    SHAMapMissingNode (SHAMapType t,
-                       SHAMapHash const& nodeHash)
-        : std::runtime_error ("SHAMapMissingNode")
-        , mType (t)
-        , mNodeHash (nodeHash)
+    SHAMapMissingNode (SHAMapType t, SHAMapHash const& hash)
```
Comment (seelabs) [inline src/ripple/shamap/SHAMapMissingNode.h:56]: What do you think of adding the stack trace here when in debug mode?

File: src/ripple/app/consensus/RCLConsensus.cpp [hunk 2]
```diff
@@ -830,7 +830,7 @@ RCLConsensus::timerEntry(NetClock::time_point const& now)
     catch (SHAMapMissingNode const& mn)
     {
         // This should never happen
-        JLOG(j_.error()) << "Missing node during consensus process " << mn;
+        JLOG(j_.error()) << "During consensus process: " << mn.what();
```
Comment (miguelportilla) [inline src/ripple/app/consensus/RCLConsensus.cpp:833]: To determine which function caught, consider adding the function name to the log.
eg.
```JLOG(j_.error()) << "During consensus process: " << mn.what() << ". Caught in " << __func__;```

File: src/ripple/app/consensus/RCLConsensus.cpp [hunk 3]
```diff
@@ -846,7 +846,7 @@ RCLConsensus::gotTxSet(NetClock::time_point const& now, RCLTxSet const& txSet)
...
```
Comment (miguelportilla) [inline src/ripple/app/consensus/RCLConsensus.cpp:849]: Consider adding the function name.
eg.
```JLOG(j_.error()) << "During consensus process: " << mn.what() << ". Caught in " << __func__;```

File: src/ripple/shamap/SHAMapMissingNode.h [hunk 4]
```diff
@@ -34,33 +36,44 @@ enum class SHAMapType
...
+to_string(SHAMapType t)
+{
+    if (t == SHAMapType::TRANSACTION)
+        return "Transaction Tree";
+    if (t == SHAMapType::STATE)
+        return "State Tree";
+    if (t == SHAMapType::FREE)
+        return "Free Tree";
+    return std::to_string(safe_cast<std::underlying_type_t<SHAMapType>>(t));
+}
```
Comment (miguelportilla) [inline src/ripple/shamap/SHAMapMissingNode.h:50]: Consider a switch statement instead.
eg.
```
switch (t)
{
case SHAMapType::TRANSACTION:
        return "Transaction Tree";
case SHAMapType::STATE:
        return "State Tree";
case SHAMapType::FREE:
        return "Free Tree";
default:
    return std::to_string(safe_cast<std::underlying_type_t<SHAMapType>>(t));
}
```

File: src/ripple/shamap/SHAMapMissingNode.h (hunk 1 above)
Comment (nbougalis) [reply src/ripple/shamap/SHAMapMissingNode.h:56]: I thought about doing this in `LogThrow` actually.

File: src/ripple/app/consensus/RCLConsensus.cpp (hunk 2 above)
Comment (nbougalis) [reply src/ripple/app/consensus/RCLConsensus.cpp:833]: I can't wait for C++20 and [`std::source_location`](https://en.cppreference.com/w/cpp/utility/source_location). :)

Comment (thejohnfreeman) [inline src/ripple/app/consensus/RCLConsensus.cpp]: Don't need the `.what()` (here and below) if you're still defining the stream operator (it still used in other places in the code).

File: src/ripple/shamap/SHAMapMissingNode.h (hunk 1 above)
Comment (HowardHinnant) [reply src/ripple/shamap/SHAMapMissingNode.h:60]: On Mac and Linux, the existing terminate handler creates a very good stack dump.  I think we used to have a custom terminate handlers and got rid of them in order to _improve_ our stack dumps.

File: src/ripple/app/consensus/RCLConsensus.cpp (hunk 2 above)
Comment (HowardHinnant) [reply src/ripple/app/consensus/RCLConsensus.cpp]: Personally I wouldn't dare to write such a proposal until I had some field experience with it.

Comment (scottschurr) [reply src/ripple/app/consensus/RCLConsensus.cpp]: I can hear the cries of ABI compatibility already.  A change like that would need to be more carefully crafted than I know how to do.

File: src/ripple/shamap/SHAMapMissingNode.h (hunk 1 above)
Comment (nbougalis) [reply src/ripple/shamap/SHAMapMissingNode.h:60]: The question is *where* do we get the usable stack dumps from? I know on Mac you can get them in the log, but what about Linux?

File: src/ripple/app/consensus/RCLConsensus.cpp (hunk 2 above)
Comment (nbougalis) [reply src/ripple/app/consensus/RCLConsensus.cpp]: I'm thinking of just removing the `<<` and printing `what()` instead. Not sure if that's better.

Comment (nbougalis) [reply src/ripple/app/consensus/RCLConsensus.cpp]: Appreciate both of your feedback. :vulcan_salute:

File: src/ripple/shamap/SHAMapMissingNode.h (hunk 1 above)
Comment (HowardHinnant) [reply src/ripple/shamap/SHAMapMissingNode.h:60]: As not-a-Linux-user, I don't know.  I thought we had that knowledge within the rippled team though.  We used to get them mailed to us whenever a production server went down (hasn't happened in awhile).

Perhaps I'm misunderstanding this thread...

File: src/ripple/app/consensus/RCLConsensus.cpp (hunk 2 above)
Comment (thejohnfreeman) [reply src/ripple/app/consensus/RCLConsensus.cpp]: I look at the stream `operator<<` as the C++ version of `__str__` in Python or `Show` in Haskell, generally considered good practice, so I like it.

Comment (nbougalis) [reply src/ripple/app/consensus/RCLConsensus.cpp]: Hmmm. That seems reasonable.

File: src/ripple/shamap/SHAMapMissingNode.h (hunk 1 above)
Comment (nbougalis) [reply src/ripple/shamap/SHAMapMissingNode.h:60]: That was caused by a script we were running on production servers ran by Ripple that would detect the crash and automatically ran `gdb` on the resulting core file. The question is what happens if there is no core file? On the Mac we do get a very pretty call stack in the logs. We'll have to research what happens on Linux.

**Pull request closed without merging**